StackOverflow when implementing a resource locator which will return the resource class itself

Hi, I've noticed that some RESTEasy tests were failing against Wildfly because of some resource classes which have resource locators that are supposed to return the resource class itself - i.e. `this`.
Of course the annotation scanner goes crazy until the stack overflow happens because it tries to process the JAX-RS resource class recursively but I was wondering whether you think that this is an issue and that the annotation scanner should be protected against this case. WDYT?

The following stack trace represents the output of a deployment failure when an attempt to deploy a WAR with such a resource:

```
at org.jboss.logging@3.4.1.Final-redhat-00001//org.jboss.logging.Logger.getLogger(Logger.java:2490)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.dataobject.DataObjectDeque.<init>(DataObjectDeque.java:38)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.<init>(OpenApiDataObjectScanner.java:138)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiDataObjectScanner.process(OpenApiDataObjectScanner.java:163)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.util.SchemaFactory.typeToSchema(SchemaFactory.java:360)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.createResponseFromJaxRsMethod(OpenApiAnnotationScanner.java:924)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsMethod(OpenApiAnnotationScanner.java:688)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.lambda$processJaxRsResourceClass$0(OpenApiAnnotationScanner.java:422)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:420)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsSubResource(OpenApiAnnotationScanner.java:511)
        at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:426)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsSubResource(OpenApiAnnotationScanner.java:511)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsResourceClass(OpenApiAnnotationScanner.java:426)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.processJaxRsSubResource(OpenApiAnnotationScanner.java:511)

...
```

I've isolated a reproducer [1] which has one method (`FruitResource::getResourceLocator()`) to demonstrate the failure.
The reproducer should fail by default. For a successful deploy the above mentioned method method should be commented.

The SmallRye OpenAPI versions that I tested this against are 1.1.21 and 1.1.22.
 
[1]
https://github.com/fabiobrz/microprofile-openapi-example/tree/subresource-locator-repro

Hi @MikeEdgar. Thanks a lot for your feedback. 
I am still looking after it. It keeps on failing in our CI runs but it passes locally. I failed to create a reproducer, as the one I made using RESTEasy resources has been deployed successfully. So, still looking for the root cause.
I investigated the stack trace as well and got to this bit:
```
Caused by: java.lang.NullPointerException
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.ParameterProcessor.setSchemaProperties(ParameterProcessor.java:637)
	at io.smallrye.openapi//io.smallrye.openapi.runtime.scanner.ParameterProcessor.getFormBodyContent(ParameterProcessor.java:605)
```
Here's where it fails, then: https://github.com/smallrye/smallrye-open-api/blob/1.1.22/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/ParameterProcessor.java#L605

--- a/core/src/main/java/io/smallrye/openapi/runtime/scanner/ResourceParameters.java
+++ b/core/src/main/java/io/smallrye/openapi/runtime/scanner/ResourceParameters.java
@@ -38,6 +38,10 @@ public class ResourceParameters {
     }
 
     public String getOperationPath() {
+        return operationPath;
+    }
+
+    public String getFullOperationPath() {
         return pathItemPath + operationPath;
     }
 

--- a/extension-jaxrs/src/main/java/io/smallrye/openapi/jaxrs/JaxRsAnnotationScanner.java
+++ b/extension-jaxrs/src/main/java/io/smallrye/openapi/jaxrs/JaxRsAnnotationScanner.java
@@ -2,7 +2,9 @@ package io.smallrye.openapi.jaxrs;
 
 import java.lang.reflect.Modifier;
 import java.util.Collection;
+import java.util.Deque;
 import java.util.HashMap;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
@@ -53,6 +55,8 @@ import io.smallrye.openapi.runtime.util.ModelUtil;
  */
 public class JaxRsAnnotationScanner extends AbstractAnnotationScanner {
     private static final String JAXRS_PACKAGE = "javax.ws.rs";
+
+    private Deque<JaxRsSubResourceLocator> subResourceStack = new LinkedList<>();
 
     @Override
     public String getName() {
@@ -319,17 +323,28 @@ public class JaxRsAnnotationScanner extends AbstractAnnotationScanner {
             return;
         }
 
+        JaxRsSubResourceLocator locator = new JaxRsSubResourceLocator(resourceClass, method);
         ClassInfo subResourceClass = context.getIndex().getClassByName(methodReturnType.name());
 
-        if (subResourceClass != null) {
-            final String originalAppPath = this.currentAppPath;
-
+        // Do not allow the same resource locator method to be used twice (sign of infinite recursion)
+        if (subResourceClass != null && !this.subResourceStack.contains(locator)) {
             Function<AnnotationInstance, Parameter> reader = t -> ParameterReader.readParameter(context, t);
 
             ResourceParameters params = ParameterProcessor.process(context.getIndex(), resourceClass, method,
                     reader, context.getExtensions());
 
-            this.currentAppPath = super.makePath(params.getOperationPath());
+            final String originalAppPath = this.currentAppPath;
+            final String subResourcePath;
+
+            if (this.subResourceStack.isEmpty()) {
+                subResourcePath = params.getFullOperationPath();
+            } else {
+                // If we are already processing a sub-resource, ignore any @Path information from the current class
+                subResourcePath = params.getOperationPath();
+            }
+
+            this.currentAppPath = super.makePath(subResourcePath);
+            this.subResourceStack.push(locator);
 
             /*
              * Combine parameters passed previously with all of those from the current resource class and
@@ -341,6 +356,7 @@ public class JaxRsAnnotationScanner extends AbstractAnnotationScanner {
                             params.getPathItemParameters(),
                             params.getOperationParameters()));
 
+            this.subResourceStack.pop();
             this.currentAppPath = originalAppPath;
         }
     }
@@ -418,7 +434,14 @@ public class JaxRsAnnotationScanner extends AbstractAnnotationScanner {
         setOperationOnPathItem(methodType, pathItem, operation);
 
         // Figure out the path for the operation.  This is a combination of the App, Resource, and Method @Path annotations
-        String path = super.makePath(params.getOperationPath());
+        final String path;
+
+        if (this.subResourceStack.isEmpty()) {
+            path = super.makePath(params.getFullOperationPath());
+        } else {
+            // When processing a sub-resource tree, ignore any @Path information from the current class
+            path = super.makePath(params.getOperationPath());
+        }
 
         // Get or create a PathItem to hold the operation
         PathItem existingPath = ModelUtil.paths(openApi).getPathItem(path);

--- a/extension-jaxrs/src/main/java/io/smallrye/openapi/jaxrs/JaxRsSubResourceLocator.java
+++ b/extension-jaxrs/src/main/java/io/smallrye/openapi/jaxrs/JaxRsSubResourceLocator.java
@@ -1,0 +1,30 @@
+package io.smallrye.openapi.jaxrs;
+
+import java.util.Objects;
+
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.MethodInfo;
+
+class JaxRsSubResourceLocator {
+    final ClassInfo clazz;
+    final MethodInfo method;
+
+    JaxRsSubResourceLocator(ClassInfo clazz, MethodInfo method) {
+        this.clazz = clazz;
+        this.method = method;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof JaxRsSubResourceLocator) {
+            JaxRsSubResourceLocator other = (JaxRsSubResourceLocator) o;
+            return Objects.equals(this.clazz, other.clazz) && Objects.equals(this.method, other.method);
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(clazz, method);
+    }
+}

--- a/extension-jaxrs/src/test/java/io/smallrye/openapi/runtime/scanner/SubresourceScanTests.java
+++ b/extension-jaxrs/src/test/java/io/smallrye/openapi/runtime/scanner/SubresourceScanTests.java
@@ -36,7 +36,7 @@ public class SubresourceScanTests extends IndexScannerTestBase {
     @Test
     public void testResteasyMultipartInput() throws IOException, JSONException {
         test("resource.subresources-with-params.json",
-                MainTestResource.class, Sub1TestResource.class, Sub2TestResource.class);
+                MainTestResource.class, Sub1TestResource.class, Sub2TestResource.class, RecursiveLocatorResource.class);
     }
 
     /***************** Test models and resources below. ***********************/
@@ -118,6 +118,26 @@ public class SubresourceScanTests extends IndexScannerTestBase {
         @Path("{subsubid}")
         public String getSub2(@PathParam("subsubid") String subsubid) {
             return null;
+        }
+    }
+
+    @Path("/recursion")
+    @SuppressWarnings("unused")
+    static class RecursiveLocatorResource {
+        @GET
+        @Path("fetch")
+        public String get() {
+            return null;
+        }
+
+        @Path("alternate1")
+        public RecursiveLocatorResource getLocator1() {
+            return this;
+        }
+
+        @Path("alternate2")
+        public RecursiveLocatorResource getLocator2() {
+            return this;
         }
     }
 }

--- a/extension-jaxrs/src/test/resources/io/smallrye/openapi/runtime/scanner/resource.subresources-with-params.json
+++ b/extension-jaxrs/src/test/resources/io/smallrye/openapi/runtime/scanner/resource.subresources-with-params.json
@@ -1,6 +1,86 @@
 {
   "openapi": "3.0.1",
   "paths": {
+    "/recursion/alternate1/alternate2/fetch": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "type": "string"
+                }
+              }
+            }
+          }
+        }
+      }
+    },
+    "/recursion/alternate1/fetch": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "type": "string"
+                }
+              }
+            }
+          }
+        }
+      }
+    },
+    "/recursion/alternate2/alternate1/fetch": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "type": "string"
+                }
+              }
+            }
+          }
+        }
+      }
+    },
+    "/recursion/alternate2/fetch": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "type": "string"
+                }
+              }
+            }
+          }
+        }
+      }
+    },
+    "/recursion/fetch": {
+      "get": {
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "type": "string"
+                }
+              }
+            }
+          }
+        }
+      }
+    },
     "/resource{resource}/sub/{id}{idMatrix}": {
       "get": {
         "parameters": [

